mlx/c/fast.h

```diff
@@ -24,53 +24,116 @@ extern "C" {
+    mlx_array* res,
+    const mlx_array w,
+    const mlx_array scales,
+    const mlx_array biases,
     int group_size,
     int bits,
-    mlx_stream s);
-mlx_array mlx_fast_affine_quantize(
-    mlx_array w,
-    mlx_array scales,
-    mlx_array biases,
+    const mlx_stream s);
+int mlx_fast_affine_quantize(
+    mlx_array* res,
+    const mlx_array w,
+    const mlx_array scales,
+    const mlx_array biases,
     int group_size,
     int bits,
-    mlx_stream s);
-mlx_array mlx_fast_layer_norm(
-    mlx_array x,
-    mlx_array weight,
-    mlx_array bias,
+    const mlx_stream s);
+int mlx_fast_layer_norm(
+    mlx_array* res,
+    const mlx_array x,
+    const mlx_array weight /* may be null */,
+    const mlx_array bias /* may be null */,
     float eps,
-    mlx_stream s);
-mlx_closure_metal_kernel_function mlx_fast_metal_kernel(
-    mlx_string name,
-    const mlx_vector_string input_names,
-    const mlx_vector_string output_names,
-    mlx_string source,
-    mlx_string header,
-    bool ensure_row_contiguous,
-    bool atomic_outputs);
-mlx_array
-mlx_fast_rms_norm(mlx_array x, mlx_array weight, float eps, mlx_stream s);
-mlx_array mlx_fast_rope(
-    mlx_array x,
+    const mlx_stream s);
+
+typedef struct mlx_fast_metal_kernel_ {
+  void* ctx;
+} mlx_fast_metal_kernel;
+mlx_fast_metal_kernel mlx_fast_metal_kernel_new(
+    const char* name,
+    const char* source,
+    const char* header);
+void mlx_fast_metal_kernel_free(mlx_fast_metal_kernel cls);
+int mlx_fast_metal_kernel_set_input_names(
+    mlx_fast_metal_kernel cls,
+    int num,
+    ...);
+int mlx_fast_metal_kernel_set_output_names(
+    mlx_fast_metal_kernel cls,
+    int num,
+    ...);
+int mlx_fast_metal_kernel_set_contiguous_rows(
+    mlx_fast_metal_kernel cls,
+    bool flag);
+int mlx_fast_metal_kernel_set_atomic_outputs(
+    mlx_fast_metal_kernel cls,
+    bool flag);
+
+int mlx_fast_metal_kernel_add_output_arg(
+    mlx_fast_metal_kernel cls,
+    const int* shape,
+    size_t size,
+    mlx_dtype dtype);
+int mlx_fast_metal_kernel_set_grid(
+    mlx_fast_metal_kernel cls,
+    int grid1,
+    int grid2,
+    int grid3);
+int mlx_fast_metal_kernel_set_thread_group(
+    mlx_fast_metal_kernel cls,
+    int thread1,
+    int thread2,
+    int thread3);
+int mlx_fast_metal_kernel_set_init_value(
+    mlx_fast_metal_kernel cls,
+    float value);
+int mlx_fast_metal_kernel_set_verbose(mlx_fast_metal_kernel cls, bool verbose);
+int mlx_fast_metal_kernel_add_template_arg_dtype(
+    mlx_fast_metal_kernel cls,
+    const char* name,
+    mlx_dtype dtype);
+int mlx_fast_metal_kernel_add_template_arg_int(
+    mlx_fast_metal_kernel cls,
+    const char* name,
+    int value);
+int mlx_fast_metal_kernel_add_template_arg_bool(
+    mlx_fast_metal_kernel cls,
+    const char* name,
+    bool value);
+
+int mlx_fast_metal_kernel_apply(
+    mlx_fast_metal_kernel cls,
+    const mlx_vector_array inputs,
+    const mlx_stream stream,
+    mlx_vector_array* outputs);
```

Comment:
Per offline discussion this function doesn't match the pattern of other functions -- the out parameter is last instead of first.

Reply:
indeed, fixed, thanks!